Issues/14909 adds new tracks event

aerych opened the pull request:
Closes #14909 

This PR adds a new event to track when the list of likers paginates (loads more likes).  This is the same event bumped by WPiOS via https://github.com/wordpress-mobile/WordPress-iOS/pull/16728. On iOS the value of source is either "notifications" or "reader", rather than "like_notification_list" or "like_reader_list". (My suggestion, I didn't realize that was an attempt to match previous values across platform 😬.)  Here I've opted to preserve the intended values.

To test:
- View a likes list, either from Notifications or the Reader.  If the reader, an en.blog post should have plenty of likers to test with.
- Scroll the list. 
- Check the log and confirm that the new event was tracked.
- Confirm the event does not repeatedly fire once you've reached the end of the list.
- Confirm the event does not fire when there is no need to paginate. 

Example: 
> 2021-06-28 16:14:16.800 8781-8781/org.wordpress.android.prealpha I/WordPress-STATS: 🔵 Tracked: like_list_fetched_more, Properties: {"list_type":"post","source":"like_notification_list"}

## Regression Notes
1. Potential unintended areas of impact
n/a

2. What I did to test those areas of impact (or what existing automated tests I relied on)
n/a

3. What automated tests I added (or what prevented me from doing so)
n/a
PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

aerych committed:
Nosara: Adds new event.

aerych committed:
Analytics: Adds method to bump event when loading more likes.

aerych committed:
Likes: Track when the list loads more.

aerych committed:
Fixes typo and lint warnings.

develric reviewed (commented):
Hey @aerych , thanks for taking care of this 🙇 ! Code LGTM and the tracking event works as described 👍 . Good job sir 😄  ! I think we are fine to go but wanted to share this thoughts I had just to be on the safe side, let me know wdyt 🙇 

- To make the data collected from the 2 platforms (iOS and Android) comparable , I was wondering how long is a page of likes in iOS (in Android it is currently set to a default of 20 likes);
- With this PR we are tracking every time we load a page (and page lenght as per previous point can affect frequency of the event). Probably this is just what we want, but was wondering if it could make sense to include in the event properties something like the page we are on (or the page we are going to request). Didn't try it but doing some `mod` operation on the number of likes got so far and the page lenght we should be able to get the data for it I think 🤔

aerych commented:
> I was wondering how long is a page of likes in iOS (in Android it is currently set to a default of 20 likes)

iOS is fetching 90 currently.  I'd like to be consistent across platforms, but I'm curious to leave the difference until the iOS feature is fully released. I know we went with 20 on Android to be cautious about performance.  Let's see how things look in a few weeks and then align the counts one way or the other.  What do ya think? 

> Probably this is just what we want, but was wondering if it could make sense to include in the event properties something like the page we are on (or the page we are going to request).

That is an excellent idea. Maybe a param for total pages also. :)  I'll add this change.

aerych committed:
Analytics: Track page number requested, and total pages available.

aerych commented:
I've pushed changes to include properties for the page requested (page) and the number of pages available (total_pages). I think the code might be a little ugly. Tips for improvements are very welcome! :) 

The logged out put should resemble the following: 
🔵 Tracked: like_list_fetched_more, Properties: {"list_type":"post","source":"like_reader_list","page":2,"total_pages":11}
🔵 Tracked: like_list_fetched_more, Properties: {"list_type":"post","source":"like_reader_list","page":3,"total_pages":11}
🔵 Tracked: like_list_fetched_more, Properties: {"list_type":"post","source":"like_reader_list","page":4,"total_pages":11}
🔵 Tracked: like_list_fetched_more, Properties: {"list_type":"post","source":"like_reader_list","page":5,"total_pages":11}
🔵 Tracked: like_list_fetched_more, Properties: {"list_type":"post","source":"like_reader_list","page":6,"total_pages":11}

Lemme know what you think!

aerych committed:
Removes unneeded scope modifier.

aerych committed:
Analytics: Passes paging information as an argument rather than computing it separately.
Updates analytics to accept new field for page length.
Hat tip @develric for the guidance. :)

aerych committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-Android into issues/14909-adds-new-tracks-event

aerych commented:
It's taken me a bit of time to loop back to this due to some AFK and other items that needed priority, but this afternoon I was able to apply @develric's suggestions to improve the way pagination information was wrangled (thank you sir!).  
(TIL about kotiln data classes. :).  Nifty! )
Ready for another look at your convenience sir. 🙇

aerych commented:
Example tracks after the latest update.  Note the inclusion of page_length.
```
🔵 Tracked: like_list_fetched_more, Properties: {"list_type":"post","page_length":20,"source":"like_reader_list","page":2,"total_pages":27}
🔵 Tracked: like_list_fetched_more, Properties: {"list_type":"post","page_length":20,"source":"like_reader_list","page":3,"total_pages":27}
🔵 Tracked: like_list_fetched_more, Properties: {"list_type":"post","page_length":20,"source":"like_reader_list","page":4,"total_pages":27}
🔵 Tracked: like_list_fetched_more, Properties: {"list_type":"post","page_length":20,"source":"like_reader_list","page":5,"total_pages":27}
```

aerych committed:
Analytics: Renames property from 'page_length' to 'per_page' for added clarity

aerych commented:
It occurred to me that `page_length` could be interpreted as the size of the page returned, rather than the size requested. I changed the property name to `per_page` as this seems a little clearer to me.

develric commented on WordPress/src/main/java/org/wordpress/android/ui/engagement/GetLikesHandler.kt:
```diff
@@ -87,7 +87,7 @@ class GetLikesHandler @Inject constructor(
 
     companion object {
-        private const val LIKES_PER_PAGE_DEFAULT = 20
+        const val LIKES_PER_PAGE_DEFAULT = 20
```
np: given the approach we have landed to with the pagingInfo data class, I think this could stay private, wdyt?

develric commented on WordPress/src/main/java/org/wordpress/android/ui/engagement/GetLikesUseCase.kt:
```diff
@@ -128,11 +129,17 @@ class GetLikesUseCase @Inject constructor(
+                val pageInfo = PagingInfo(
+                        pageLength = paginationParams.pageLength,
+                        page = if (likes.isNotEmpty()) ((likes.size - 1) / paginationParams.pageLength) + 1 else 1,
+                        totalPages = ceil(fingerPrint.expectedNumLikes.toDouble() / paginationParams.pageLength).toInt()
```
np: to guard from having the paginationParams.pageLength possibly equal to 0, maybe we can consider adding the following

```
require(pageLength != 0) { "The page length for likes cannot be 0" }
```

both at the beginning of `suspend fun handleGetLikesForPost(` and `suspend fun handleGetLikesForComment(` in the `GetLikesHandler`

aerych replied:
I'm not sure I understand this one. It seems like this would yield a fatal exception if a zero was ever encountered? Would this be automatically handled by an existing mechanism or would it need some additional handling and logging around it.   Wondering if it should be out of scope here since we're focused on adding analytics?  What do you think?

aerych replied:
Thanks for clarifying in slack sir!  I've added the changes and included a comment with some of the rationale.

develric commented on WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsUtils.java:
```diff
@@ -75,6 +75,9 @@
     private static final String IS_STORAGE_SETTINGS_RESOLVED_KEY = "is_storage_settings_resolved";
+    private static final String PAGE_KEY = "page";
+    private static final String TOTAL_PAGES_KEY = "total_pages";
+    private static final String PAGE_LENGTH_KEY = "per_page";
```
yep `per_page` is definitely clearer 👍 !

develric reviewed (commented):
Thanks for the follow up @aerych , I was finally able to further look into this one and overall I think we are pretty there 👍 .

I think it's working great for reader post likes (where as a side note, the `fingerPrint.expectedNumLikes` is populated from the post data because we need that info to create the likes list title), but we have an issue when tracking from notifications since the `fingerPrint.expectedNumLikes` is not populated there because AFAIU the notifications endpoint is not giving us that info at hand. 

I would need to better look into it, but If we want to keep the totalPages info, one possibility can be to get that information from the likes endpoint (the `found` field), and a mapping in FluxC should be needed in that case. We will store the found value in the cache (and do not update it during pagination after the first successful fetch) so that we can use it for cases when an error is returned from the network fetching.

If (and it's pretty possible, you know my English sir 🙃 ) the previous does not make a clear sense, just ping me and we can align on how to better proceed 🙇

aerych committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-Android into issues/14909-adds-new-tracks-event

aerych commented:
> If we want to keep the totalPages info,

Honestly, I think it's just a "nice to have" and not something we _need_ to have.  Since it's not readily supported by both endpoints let's drop that part for now. I'll make the changes. :)

aerych committed:
Analytics: Addressing feedback.

aerych committed:
Tests: Updates tests with stubbed PageInfo

aerych commented:
Heya @develric 👋   I've made a few changes based on feedback. `LIKES_PER_PAGE_DEFAULT` is again private and I've changed the `pageLength` argument in the analytics methods to be `perPage` to better match the earlier change.  I've left other instances of `pageLength` alone. 
I've also removed the code related to tracking total pages to keep things simple (we probably don't need that property anyway).
Finally, I've updated unit tests with a stubbed PageInfo property where needed.  Lemme know what you think. 
Ready for another look!

aerych committed:
Remove unneeded import.

aerych committed:
Adds require statements as a safety net.

develric reviewed (approved):
Thanks for the changes and for taking care of this @aerych 🙇 ! Changes LGTM and made another test round that was ok 👍 ! :shipit:

aerych commented:
Yay! Thanks for all your help Riccardo! 🙇
